File: tic_tac_toe.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <string>
#include <algorithm>
bool  is_filled(const std::vector<std::string> board){
    for (const auto &row : board) {
        if (std::any_of(row.begin(), row.end(), [](const char &c) {
            return c == '-';
        })) {
            return false;
        }
    }
    return true;
}
int are_equal(const std::string &row){
    char c = row.front();
    return std::all_of(row.begin(), row.end(), [&c](const char &a) {
        return  a != '-' && a == c;
    });
}
int are_equal(const char &a, const char &b, const char &c){
    if(a==b && a==c && a!= '-'){
        return 1;
    }
    return 0;
}
class tic_tac_toe{
   private:
    struct Player {
        std::string name;
        char sign;
    };
    // char  play_board[3][3];
    // std::vector<std::vector<char>> play_board(3, std::vector<char>(3, '-'));
    std::vector<Player> players;
    std::vector<std::string> play_board;
    public:
    tic_tac_toe() : play_board(std::vector<std::string>(3, "---")), players(std::initializer_list<Player>{Player{"", 'X'}, Player{"", 'O'}}){}
    void display();
    void reset();
    std::pair<bool, int> insert(const Player &);
    void master();
    void get_name();
    bool game_cond(const int &);
    void start();
};
void tic_tac_toe::start(){
   std::system("clear");
   while(1){
    
std::cout <<"   ▄      ▀                    ▄                           ▄                 \n" 
          << " ▄▄█▄▄  ▄▄▄     ▄▄▄          ▄▄█▄▄   ▄▄▄    ▄▄▄          ▄▄█▄▄   ▄▄▄    ▄▄▄  \n"
          << "   █      █    █▀  ▀           █    ▀   █  █▀  ▀           █    █▀ ▀█  █▀  █ \n"
          << "   █      █    █       ▀▀▀     █    ▄▀▀▀█  █       ▀▀▀     █    █   █  █▀▀▀▀ \n"
          << "   ▀▄▄  ▄▄█▄▄  ▀█▄▄▀           ▀▄▄  ▀▄▄▀█  ▀█▄▄▀           ▀▄▄  ▀█▄█▀  ▀█▄▄▀ \n\n";                                                                          
                                                                        
    std::string choice;
    std::cout<<"1. New game\n2. Quit"<<std::endl;
    do{
        std::cin>>choice;
    if (choice.size() > 1) {
        std::cout << "Invalid input! try again\n";
        continue;
    }
    if ( choice.front() == '2') {
        return;
    }
    else if (choice.front() == '1') {
        master();
        break;
    }
    else {
        std::cout << "Invalid input! try again\n";
        continue;
    }
    }while(true);
   }
}
bool tic_tac_toe::game_cond(const int &idx){
    int row = idx / 3, col = idx % 3;
    if (are_equal(play_board[row])) {
        return true;
    }
    if (are_equal(play_board[0][col], play_board[1][col], play_board[2][col])) {
        return true;
    }
    if (idx == 4 && 
    (are_equal(play_board[0][0], play_board[1][1], play_board[2][2]) ||
     are_equal(play_board[0][2], play_board[1][1], play_board[2][0]))
    ) {
       return true; 
    }
    if ((idx == 0 || idx == 8) && are_equal(play_board[0][0], play_board[1][1], play_board[2][2])) {
        return true;
    }
    if ((idx == 2 || idx == 6) && are_equal(play_board[0][2], play_board[1][1], play_board[2][0])) {
        return true;
    }
    return false;
}
void tic_tac_toe::get_name(){
    std::cout<<"Name of player 1: ";
    std::cin>>players[0].name;
    std::cout<<"Name of player 2: ";
    std::cin>>players[1].name;
    std::cout<<std::endl;
}
std::pair<bool, int> tic_tac_toe::insert(const Player &player){
    std::string str;
    std::cout<<player.name<<"'s turn: ";
    std::cin>>str;
    char val = str.front();
    if (val == 'q' || val == 'Q') {
        return {false, 0};
    }
    if (str.size() != 1 || val < '1' || val > '9') {
        std::cout << "Invalid input, try again" << std::endl;
        return insert(player);
    }
    else {
        int idx = val - '0' - 1;
        char &pos = play_board[idx / 3][idx % 3];
        if (pos != '-') {
            std::cout<<"Error! Already occupied! try again"<<std::endl;
            insert(player);
        }
        else {
            pos = player.sign;
        }
        return {true, idx};
    }
    return {true, 0};
}
void tic_tac_toe::master(){
    std::system("clear");
    reset();
    get_name();
    std::cout<<"Instruction: Enter the index of the position[1 - 9]"<<std::endl;
    int player_idx = 0;
    while(true){
        display();
        auto cond = insert(players[player_idx]);
        if(!cond.first) {
            std::system("clear");
            return; 
        }
        if (game_cond(cond.second)) {
            std::system("clear");
            std::cout<<"\033[3;31m"<<players[player_idx].name<<" Wins!\n\n\033[0m";
            return;
        }
        if(is_filled(play_board)){
            std::system("clear");
            std::cout<<"It ended in a draw!\n\n";
            return;
        }
        player_idx = player_idx ? 0 : 1;
    }
}
void tic_tac_toe::display(){
    std::system("clear");
    std::cout << "Enter a number between 1 to 9 (Enter q to quit the game)\n\n";
    for(int i=0;i<3;i++){
        // std::cout<<std::setw(20);
        for(int j=0;j<3;j++){
            std::cout<<play_board[i][j]<<std::setw(6);
        }
        std::cout<<"\n\n";
    }
    
}

void tic_tac_toe::reset(){
    for (auto &row : play_board) {
        row = "---";
    }
}

int main(){
    tic_tac_toe game;
    game.start();
    std::system("clear");
    return 0;
}

File: tic_tac_toe.py
import os
import curses


class tic_tac_toe:
    def __init__(self) -> None:
        self.players = [{"name" : "", "sign" : 'X'}, {"name" : "", "sign" : 'O'}]
        self.play_board = ["---", "---", "---"]
        self.option = [int(0), int(0)]

    def start(self, stdscr):
        os.system("clear")
        options = ["| New Game", "| Quit"]
        while (True):
            attributes = {}
            curses.init_pair(1, curses.COLOR_WHITE, curses.COLOR_BLACK)
            attributes['normal'] = curses.color_pair(1)
            curses.init_pair(2, curses.COLOR_BLACK, curses.COLOR_WHITE)
            attributes['highlighted'] = curses.color_pair(2)
            
            c = 0
            option = 0
            while c != 10:
                stdscr.addstr("   ▄      ▀                    ▄                           ▄                 \n")
                stdscr.addstr(" ▄▄█▄▄  ▄▄▄     ▄▄▄          ▄▄█▄▄   ▄▄▄    ▄▄▄          ▄▄█▄▄   ▄▄▄    ▄▄▄  \n")
                stdscr.addstr("   █      █    █▀  ▀           █    ▀   █  █▀  ▀           █    █▀ ▀█  █▀  █ \n")
                stdscr.addstr("   █      █    █       ▀▀▀     █    ▄▀▀▀█  █       ▀▀▀     █    █   █  █▀▀▀▀ \n")
                stdscr.addstr("   ▀▄▄  ▄▄█▄▄  ▀█▄▄▀           ▀▄▄  ▀▄▄▀█  ▀█▄▄▀           ▀▄▄  ▀█▄█▀  ▀█▄▄▀ \n\n")                                                                          
            
                for i in range(0, 2):
                    if i == option:
                        attr = attributes['highlighted']
                    else:
                        attr = attributes['normal']
                    stdscr.addstr(options[i] + '\n', attr | curses.A_BOLD)
                c = stdscr.getch()
                if c == curses.KEY_UP and option > 0:
                    option -= 1
                elif c == curses.KEY_DOWN and option < len(options) - 1:
                    option += 1
                stdscr.erase()
            
            if (option == 0):
               self.master(stdscr) 
            else:
                return






    def reset(self):
        self.play_board = ["---", "---", "---"]
                


    def is_filled(self, stdscr)->bool:
        for row in self.play_board:
            c = row.find('-')
            if (row.find('-') != int(-1)):
                return False
        return True

    def insert(self, stdscr, player):
       while (True):
                
            attributes = {}
            curses.init_pair(1, curses.COLOR_WHITE, curses.COLOR_BLACK)
            attributes['normal'] = curses.color_pair(1)
            curses.init_pair(2, curses.COLOR_BLACK, curses.COLOR_WHITE)
            attributes['highlighted'] = curses.color_pair(2)
                
            c = 0
            while c != 10:
                if (c == 'q' or c == 'Q'):
                    return [False, 0]
                stdscr.erase()
                stdscr.addstr("Player " + str(player + 1)+ "'s turn:\n\n", curses.A_BOLD)
                for i in range(0, 3):
                    for j in range(0, 3):
                            if ([i, j] == self.option):
                                attr = attributes['highlighted']
                            else:
                                attr = attributes['normal']
                            stdscr.addstr(" " + self.play_board[i][j] + " ", attr | curses.A_BOLD)
                            stdscr.addstr("  ")
                    stdscr.addstr("\n\n")
                c = stdscr.getch()
                if c == curses.KEY_UP and self.option[0] > 0:
                        self.option[0] -= 1
                elif c == curses.KEY_DOWN and self.option[0] < 2:
                    self.option[0] += 1
                elif c == curses.KEY_LEFT and self.option[1] > 0:
                        self.option[1] -= 1
                elif c == curses.KEY_RIGHT and self.option[1] < 2:
                    self.option[1] += 1
            if (self.play_board[self.option[0]][self.option[1]] != '-'):
                continue
            else:
                if (player == 0):
                    list1 = list(self.play_board[self.option[0]])
                    list1[self.option[1]] = 'X'
                    self.play_board[self.option[0]] = ''.join(list1)
                else:
                    list1 = list(self.play_board[self.option[0]])
                    list1[self.option[1]] = '0'
                    self.play_board[self.option[0]] = ''.join(list1)
                return [True, self.option[0] * 3 + self.option[1]]
        
            


    def master(self, stdscr):
        self.reset()
        stdscr.clear()
        # self.get_name(stdscr)
        player_idx = int(0)
        while (True):
            curr_cond = self.insert(stdscr, player_idx)
            if (curr_cond[0] is not True):
                stdscr.erase()
                return
            if (self.game_cond(int(curr_cond[1])) is True):
                stdscr.erase()
                stdscr.addstr("Player " + str(player_idx + 1) + " Wins!\n\n", curses.A_ITALIC | curses.A_BOLD  )
                return
            if (self.is_filled(stdscr) is True):
                stdscr.erase()
                stdscr.addstr("It ended in a draw!\n\n", curses.A_ITALIC | curses.A_BOLD)
                return
            player_idx = (player_idx + 1) % 2


    def get_name(self, stdscr):
        stdscr.addstr("Name of Player 1: ")
        self.players[0]["name"] = stdscr.getstr()
        stdscr.addstr("Name of Player 2: ")
        self.players[1]["name"] = stdscr.getstr()
    
    def game_cond(self, index):
       row = int(index / 3)
       col = int(index % 3)
       if (are_equal(self.play_board[row][0], self.play_board[row][1], self.play_board[row][2])):
           return True

       if (are_equal(self.play_board[0][col], self.play_board[1][col], self.play_board[2][col])):
           return True

       if (are_equal(self.play_board[0][0], self.play_board[1][1], self.play_board[2][2])):
           return True
        
       if (are_equal(self.play_board[0][2], self.play_board[1][1], self.play_board[2][0])):
           return True

       return False


def are_equal(val1, val2, val3):
    return (val1 == val2) and (val1 == val3) and (val1 != '-')


def main(stdscr):
    game = tic_tac_toe()
    game.start(stdscr)
    stdscr.erase()
    stdscr.addstr("Press any key to exit", curses.A_ITALIC | curses.A_BOLD)
    stdscr.getch()
    


curses.wrapper(main)
